Improved documentation (#233)
NumericalMathematics/DispersiveShallowWater.jl: docs/src/SBP_Operators.md

The framework has since been extended to encompass a much broader class of methods, making it a unifying principle in numerical analysis. This generality allows practitioners to choose the most appropriate discretization for their specific problem while maintaining the same theoretical guarantees and structure-preserving properties.

[^LampertRanocha2024]:
    Lampert, Ranocha (2024):
    Structure-Preserving Numerical Methods for Two Nonlinear Systems of Dispersive Wave Equations
    [arXiv: 2402.16669](https://arxiv.org/abs/2402.16669)

## 2. Mathematical Foundation

### The Core SBP Property

The fundamental property that defines a first-derivative SBP operator is:

```math
MD + D^T M = t_R t_R^T - t_L t_L^T
```

where:

- ``D`` is the derivative operator (matrix)
- ``M`` is the symmetric, positive definite mass matrix

> - ``t_L = (1,0,...,0)^T`` and ``t_R = (0,...,0,1)^T`` extract boundary values

Reviewer: ```suggestion
- ``\boldsymbol{e}_L = (1,0,...,0)^T`` and ``\boldsymbol{e}_R = (0,...,0,1)^T`` extract boundary values
```
Could we use ``\boldsymbol{e}_L/R`` for these unit vectors to stay consistent with the paper and also make them boldface since they are vectors and below we use boldface to indicate the variable is a vector.

This property is the discrete analog of integration by parts:

```math
\begin{aligned}

> \underbrace{ \boldsymbol{u}^T M D \boldsymbol{v} + \boldsymbol{u}^T D^T M \boldsymbol{v} }_{\displaystyle \approx \int_{x_{\min}}^{x_{\max}} u\, (\partial_x v) + \int_{x_{\min}}^{x_{\max}} (\partial_x u)\, v } 

Reviewer: ```suggestion
\underbrace{ \boldsymbol{u}^T M D \boldsymbol{v} + \boldsymbol{u}^T D^T M \boldsymbol{v} }_{\displaystyle \approx \int_{x_{\min}}^{x_{\max}} u\, (\partial_x v)\textrm{d}x + \int_{x_{\min}}^{x_{\max}} (\partial_x u)\, v\textrm{d}x } 
```
Later, we also use the `dx` behind the integral.

&= 

> \underbrace{ \boldsymbol{u}^T \boldsymbol{t}_R \boldsymbol{t}_R^T \boldsymbol{v} - \boldsymbol{u}^T \boldsymbol{t}_L \boldsymbol{t}_L^T \boldsymbol{v} }_{\displaystyle \approx u(x_{\max})\,v(x_{\max}) - u(x_{\min})\,v(x_{\min}) }.

Reviewer: ```suggestion
\underbrace{ \boldsymbol{u}^T \boldsymbol{t}_R \boldsymbol{t}_R^T \boldsymbol{v} - \boldsymbol{u}^T \boldsymbol{t}_L \boldsymbol{t}_L^T \boldsymbol{v} }_{\displaystyle = u(x_{\max})\,v(x_{\max}) - u(x_{\min})\,v(x_{\min}) }.
```
These are equal, right?

\end{aligned}

> ```

Reviewer: I would like to highlight that the two terms under the braces are equal (integration by parts), which is not quite clear from the current equation. Maybe you could reuse the following LaTeX code (or take it as an inspiration):
```tex
\begin{equation*}
        \begin{array}{ccccccc}
            \underbrace{\bm u^TMD_1\bm v}_{\approx} &+& \underbrace{\bm u^TD_1^TM\bm v}_{\approx} &=& \underbrace{\bm u^T\bm e_R\bm e_R^T\bm v}_{\eq} &-& \underbrace{\bm u^T\bm e_L\bm e_L^T\bm v}_{\eq}\\
            \overbrace{\displaystyle\int_\Omega uv_x\tr dx} &+& \overbrace{\displaystyle\int_\Omega u_xv\tr dx} &=& \overbrace{u(x_{\text{max}})v(x_{\text{max}})} &-& \overbrace{u(x_{\text{min}})v(x_{\text{min}})}.
        \end{array}
    \end{equation*}
```
(probably you need to replace `\bm` by `\boldsymbol` and `\tr` by `\textrm`.

where for periodic SBP operators their property naturally simplifies to

```math
MD + D^T M = 0.
```

### Understanding the Mass Matrix

The mass matrix ``M`` approximates the continuous inner product:

```math

> \langle u, v \rangle_M = u^T M v \approx \int_{x_{min}}^{x_{max}} u(x) v(x) dx

Reviewer: ```suggestion
\langle \boldsymbol{u}, \boldsymbol{v} \rangle_M = \boldsymbol{u}^T M \boldsymbol{v} \approx \int_{x_{min}}^{x_{max}} u(x) v(x) dx
```
On the left-hand side we are dealing with vectors, on the right-hand side with functions.

```

For the approximation to be meaningful, we require:

```math

> 1^T M 1 = x_{max} - x_{min}

Reviewer: ```suggestion
\boldsymbol{1}^T M \boldsymbol{1} = x_{\max} - x_{\min}
```
I'm not sure if we need to mention this explicitly. Actually this simply follows if D is at lest first order (which we usually have):

$$
\boldsymbol{1}^T M \boldsymbol{1} = \boldsymbol{1}^T M D\boldsymbol{x} = -\boldsymbol{1}^T D^T M\boldsymbol{x} + \boldsymbol{1}^T(\boldsymbol{e}_R \boldsymbol{e}_R^T - \boldsymbol{e}_L \boldsymbol{e}_L^T)\boldsymbol{x} = -\cancel{(D \boldsymbol{1})^T} M \boldsymbol{x} + x_N - x_1 = x\_{\max} - x\_{\min}
$$

Author: if i remember correctly hendrik did also explicitly mention it in our num pde script.

```

This ensures that the discrete inner product correctly integrates constants.

### Understanding the Derivative Operator

The derivative operator ``D`` approximates the spatial derivative of a function on a discrete grid. That is, for a smooth function ``u(x)``, we want the discrete expression ``D \boldsymbol{u}`` to approximate ``\partial_x u(x)`` at the grid points.

A first-derivative SBP operator is said to be **$p$-th order accurate** if it satisfies:

```math
D\, \boldsymbol{x}^k = k\, \boldsymbol{x}^{k-1}, \quad \text{for all } k = 0, 1, ..., p,
```

> where ``\boldsymbol{x}^k = (x_1^k, \dots, x_N^k)^T`` and the operations are performed pointwise. This means the discrete operator correctly differentiates monomials up to degree $p$. In particular, **consistency** requires at least zeroth-order accuracy, i.e.,

Reviewer: ```suggestion
where ``\boldsymbol{x}^k = (x_1^k, \dots, x_N^k)^T``. This means the discrete operator correctly differentiates monomials up to degree $p$. In particular, **consistency** requires at least zeroth-order accuracy, i.e.,
```
I think the only pointwise operation is the exponentiation, but if you already explicitly define that, we don't need to say again that the operation is performed pointwise (and otherwise, one could be confused with $$D\boldsymbol{x}$$ being performed pointwise.

```math
D\, \boldsymbol{1} = \boldsymbol{0}.
```

> However, approximating derivatives near boundaries poses a challenge: interior points can use standard finite difference stencils, but near the edges (e.g., at `x_1` or `x_N`), one must use specially designed one-sided approximations that still preserve stability and accuracy. Periodic SBP Operators done have this problem.

Reviewer: ```suggestion
However, approximating derivatives near boundaries poses a challenge: interior points can use standard finite difference stencils, but near the edges (e.g., at `x_1` or `x_N`), one must use specially designed one-sided approximations that still preserve stability and accuracy. Periodic SBP Operators do not have this problem.
```

### Example: Central Difference Operator

A simple example is the classical second-order central difference operator on a uniform grid:

```math
D = \frac{1}{\Delta x} \begin{pmatrix}
-1 & 1 & & & \\
-1/2 & 0 & 1/2 & & \\
& -1/2 & 0 & 1/2 & \\
& & \ddots & \ddots & \ddots \\
& & & -1 & 1
\end{pmatrix}
```

with mass matrix:

```math
M = \Delta x \cdot \text{diag}(1/2, 1, 1, \ldots, 1, 1/2)
```

You can verify that this satisfies the SBP property and provides second-order accuracy in the interior with first-order accuracy at the boundaries.

## 3. Types of SBP Operators

> In practice SBP Operators come in all kind of *flavours*. 

Reviewer: ```suggestion
In practice SBP operators come in various *flavours*. 
```

### Example: Central Difference Operator

A simple example is the classical second-order central difference operator on a uniform grid:

```math
D = \frac{1}{\Delta x} \begin{pmatrix}
-1 & 1 & & & \\
-1/2 & 0 & 1/2 & & \\
& -1/2 & 0 & 1/2 & \\
& & \ddots & \ddots & \ddots \\
& & & -1 & 1
\end{pmatrix}
```

with mass matrix:

```math
M = \Delta x \cdot \text{diag}(1/2, 1, 1, \ldots, 1, 1/2)
```

You can verify that this satisfies the SBP property and provides second-order accuracy in the interior with first-order accuracy at the boundaries.

## 3. Types of SBP Operators

In practice SBP operators come in various *flavors*. Consult the [documentation of SummationByPartsOperators.jl](https://ranocha.de/SummationByPartsOperators.jl/stable/)
for more details on how to construct the different types.

### Central Finite Difference (FD) SBP Operators

The simplest form of SBP operators are central finite difference operators and these are the operators that are mostly used in DispersiveShallowWater.jl
For periodic boundary conditions, they can be created with `periodic_derivative_operator` and for non-periodic boundary conditions with `derivative_operator`

> from SummationByPartsOperators.jl.

Reviewer 2: ```suggestion
from [SummationByPartsOperators.jl](https://github.com/ranocha/SummationByPartsOperators.jl/).
```

### [Upwind SBP Operators](@id upwind_sbp)

Upwind operators come in pairs ``D_+`` and ``D_-``, which satisfy the relation

```math
MD_+ + D_-^T M = \boldsymbol{e}_R \boldsymbol{e}_R^T - \boldsymbol{e}_L \boldsymbol{e}_L^T,
```

where additionally ``M(D_+ - D_-)`` is negative semidefinite. This property is often useful to construct dissipative numerical schemes, which have an improved stability.
The operators ``D_+`` and ``D_-`` are biased in one direction and are therefore useful in simulations with unidirectional flow, which favor a specific flow direction.
They can also be helpful to construct (central) second-derivative operators by ``D_2 = D_+D_-`` (or ``D_2 = D_-D_+``), which is often advantageous compared to wide-stencil
central FD operators like ``D_2 = D_1^2`` because second-derivative operators based on upwind operators have a narrow stencil leading to a better resolution.

> With SummationByPartsOperators.jl, upwind operators can be constructed using `upwind_operators`. Each part of the operator can be accessed by the `minus` and `plus` fields.

Reviewer 2: ```suggestion
With [SummationByPartsOperators.jl](https://github.com/ranocha/SummationByPartsOperators.jl/), upwind operators can be constructed using `upwind_operators`. Each part of the operator can be accessed by the `minus` and `plus` fields.
```

D_legendre = legendre_derivative_operator(-1.0, 1.0, p + 1)
uniform_mesh = UniformPeriodicMesh1D(coordinates_min, coordinates_max, div(N, p + 1))
D = couple_discontinuously(D_legendre, uniform_mesh)
```

Using `couple_discontinuously`, you can also construct upwind SBP operators by additionally passing `Val(:plus)` or `Val(:minus)`. Note that this construction results into non-uniformly
distributed nodes and due to the discontinuous nature to repeated nodes at interfaces between elements.

### [Continuous Galerkin (CG) SBP](@id cg_sbp)

Similarly to DG operators, CG operators can be constructed using `couple_continuously`. In contrast to the DG operators, `N` needs to be divisible by `p` and there are no repeated nodes at the interfaces.

### [Fourier/Spectral SBP](@id fourier_sbp)

Fourier or spectral SBP operators are constructed using Fourier basis functions. These operators can be used for problems with periodic boundary conditions. The key idea is to represent the solution in
terms of its Fourier coefficients and to apply differentiation in the Fourier space. Note that these operators have dense derivative matrices and are therefore often more computationally expensive. In

> SummationByPartsOperators.jl, they can be constructed with `fourier_derivative_matrix`.

Reviewer 2: ```suggestion
[SummationByPartsOperators.jl](https://github.com/ranocha/SummationByPartsOperators.jl/), they can be constructed with `fourier_derivative_matrix`.
```
